gtsam/nonlinear/NonlinearFactor.h

```diff
@@ -29,11 +29,35 @@
 #include <gtsam/base/utilities.h>  // boost::index_sequence
 
 #include <boost/serialization/base_object.hpp>
+#include <cstddef>
+#include <type_traits>
 
 namespace gtsam {
 
 /* ************************************************************************* */
 
+/** These typedefs and aliases will help with making the evaluateError interface
+ * independent of boost
+ * TODO(kartikarcot): Change this to OptionalMatrixNone
+ * This typedef is used to indicate that the Jacobian is not required
+ * and the default value used for optional matrix pointer arguments in evaluateError.
+ * Had to use the static_cast of a nullptr, because the compiler is not able to
+ * deduce the type of the nullptr when expanding the evaluateError templates.
+ */
+#define OptionalNone static_cast<Matrix*>(nullptr)
+
+/** This typedef will be used everywhere boost::optional<Matrix&> reference was used
+ * previously. This is used to indicate that the Jacobian is optional. In the future
+ * we will change this to OptionalJacobian
+ */
+using OptionalMatrixType = Matrix*;
```

Comment:
Why raw pointers? Isn't `std::optional<std::reference_wrapper<T>>` working?

@jlblancoc

Reply:
It's up to you all, library API designers, both policies may make sense, despite my personal preference for not using pointers.

PS: Ownership and lifetime is neither ensured with the `std::optional<...>` type, btw... The only way to enforce and clarify both are (AFIK) smart pointers, so, dynamic memory, killing performance... I guess you must rely on well-designed code, validate with sanitizers, etc. Just as done so far :-)